modules/search/class.jetpack-search.php

```diff
@@ -0,0 +1,1289 @@
+<?php
+
+class Jetpack_Search {
+
+	protected $found_posts = 0;
+
+	/**
+	 * The maximum offset ('from' param), since deep pages get exponentially slower.
+	 *
+	 * @see https://www.elastic.co/guide/en/elasticsearch/guide/current/pagination.html
+	 */
+	protected $max_offset = 200;
+
+	protected $search_result;
+
+	protected $original_blog_id;
+	protected $jetpack_blog_id;
+
+	protected $aggregations = array();
+	protected $max_aggregations_count = 100;
+
+	protected static $instance;
+
+	//Languages with custom analyzers, other languages are supported,
+	// but are analyzed with the default analyzer.
+	public static $analyzed_langs = array( 'ar', 'bg', 'ca', 'cs', 'da', 'de', 'el', 'en', 'es', 'eu', 'fa', 'fi', 'fr', 'he', 'hi', 'hu', 'hy', 'id', 'it', 'ja', 'ko', 'nl', 'no', 'pt', 'ro', 'ru', 'sv', 'tr', 'zh' );
+
+	protected function __construct() {
+		/* Don't do anything, needs to be initialized via instance() method */
+	}
+
+	public function __clone() {
+		wp_die( "Please don't __clone Jetpack_Search" );
+	}
+
+	public function __wakeup() {
+		wp_die( "Please don't __wakeup Jetpack_Search" );
+	}
+
+	/**
+	 * Get singleton instance of Jetpack_Search
+	 *
+	 * Instantiates and sets up a new instance if needed, or returns the singleton
+	 *
+	 * @module search
+	 *
+	 * @return Jetpack_Search The Jetpack_Search singleton
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Jetpack_Search();
+
+			self::$instance->setup();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Perform various setup tasks for the class
+	 *
+	 * Checks various pre-requisites and adds hooks
+	 *
+	 * @module search
+	 */
+	public function setup() {
```

Comment:
Yeah I've thought about that some - it's not _too_ much of a problem at this point because the module must be enabled manually and search will fall back to regular posts if the API returns an error, but I think it's definitely something we need to make more robust.

I wonder if we should implement this safeguard when we make a wider rollout of this (UI, auto index creation, etc). I'm ok with either way depending on how important we feel it is.